src/typechecker/define-instance.lisp

```diff
@@ -28,17 +29,52 @@
 
 (in-package #:coalton-impl/typechecker/define-instance)
 
+(defun expand-constraint (base-constraint env)
+  "This was implemented as a hack to make `derive' work on recursive
+types.  It gets rid of recursive constraints in class instances, allowing you
+to write an instance with signature `(Eq A => Eq A)'."
+  (declare (type tc:ty-predicate base-constraint)
+           (type tc:environment env)
+           (values tc:ty-predicate-list &optional))
+
+  (labels ((f (constraint env)
+             (multiple-value-bind (inst subs)
+                 (tc:lookup-class-instance env constraint :no-error t)
+               (if inst
+                   (mapcan (a:rcurry #'f env)
```

Comment:
`mapcan f` and `mapcar g` can be combined into one function:

    (mapcan f (mapcar g xs))

    ==

    (mapcan (compose f g) xs)

The composition of f and g can be written in a lambda to eliminate the need for the curries.